akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/TestKitStorage.scala

```diff
@@ -47,15 +51,26 @@ sealed trait InMemStorage[K, R] extends InternalReprSupport[R] {
           Some(value.drop(fromInclusive).take(maxNum))
         else None)
 
+  def removeFirst(key: K): Unit = Try(eventsLog.get(key)).foreach(_.poll())
+
+  def first(key: K): Option[R] = Try(eventsLog.get(key)).toOption.flatMap { item =>
+    Try(item.element()).toOption.map(toRepr)
+  }
+
   def findOneByIndex(key: K, index: Int): Option[R] =
     Option(eventsMap.get(key))
       .flatMap {
         case (_, value) => if (value.size > index) Some(value(index)) else None
       }
       .map(toRepr)
 
-  def add(key: K, p: R): Unit =
+  def add(key: K, p: R): Unit = {
+    if (!Try(eventsLog.containsKey(key)).getOrElse(false)) {
+      eventsLog.put(key, new util.LinkedList[InternalRepr]())
```

Comment:
This can have concurrent access so to make it thread safe you should use `eventsLog.putIfAbsent` instead of the two operations containsKey + put